src/content/services/monetizationLinkManager.ts

```diff
@@ -507,60 +518,38 @@ export class MonetizationLinkManager {
     void this.sendStartMonetization(startMonetizationPayload);
   }
 
-  private async onAddedNode(
-    node: Node,
-  ): Promise<StartMonetizationPayload | null> {
-    const { HTMLElement, HTMLLinkElement } = this.global;
-
-    if (node instanceof HTMLElement) {
-      this.dispatchOnMonetizationAttrChangedEvent(node);
-    }
-
-    if (node instanceof HTMLLinkElement) {
-      const payloadEntry = await this.onAddedLink(node);
-      return payloadEntry ? [payloadEntry] : null;
-    } else if (node instanceof HTMLElement) {
-      const linkElements = this.getMonetizationLinkTags(node);
-      return await Promise.all(
-        linkElements.map((linkElem) => this.onAddedLink(linkElem)),
-      ).then((res) => res.filter(isNotNull));
+  private async onAddedLink(
+    link: HTMLLinkElement,
+  ): Promise<StartMonetizationPayloadEntry | null> {
+    if (
+      this.pendingValidationLinks.has(link) ||
+      this.monetizationLinks.has(link)
+    ) {
+      return null;
     }
-    return null;
-  }
 
-  private onRemovedNode(node: Node): StopMonetizationPayload | null {
-    const { HTMLElement, HTMLLinkElement } = this.global;
+    this.pendingValidationLinks.add(link);
 
-    if (node instanceof HTMLLinkElement) {
-      return [this.onRemovedLink(node)];
-    } else if (node instanceof HTMLElement) {
-      const linkElements = this.getMonetizationLinkTags(node).filter((el) =>
-        this.monetizationLinks.has(el),
-      );
-      return linkElements.map((linkElem) => this.onRemovedLink(linkElem));
+    const walletAddress = await this.checkLink(link);
+    if (!walletAddress) {
+      return null;
     }
 
-    return null;
-  }
-
-  private async onAddedLink(
-    link: HTMLLinkElement,
-  ): Promise<StartMonetizationPayloadEntry | null> {
+    this.monetizationLinks.set(link, walletAddress);
+    // if link validation failed, do not remove it from pending validation links
+    this.pendingValidationLinks.delete(link);
     this.observeLinkAttrs(link);
-    const res = await this.checkLink(link);
-    if (!res) return null;
-    this.monetizationLinks.set(link, res.details);
-    return res.details;
+
+    return walletAddress;
   }
 
   private onRemovedLink(link: HTMLLinkElement): StopMonetizationPayloadEntry {
     const details = this.monetizationLinks.get(link);
     if (!details) {
       throw new Error(
-        `Could not find details for monetization node ${link.outerHTML.slice(0, 200)}`,
+        `Could not find wallet address for monetization node ${link.outerHTML.slice(0, 200)}`,
```

Comment:
This could confuse us. "find wallet address" could mean `GET` response, not the one that this is an unexpected error (a sanity check).
```suggestion
        `Unexpected: No details found for monetization node ${link.outerHTML.slice(0, 200)}`,
```